### PR TITLE
bpo-40275: Avoid importing logging in test.support

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Import `logging` lazily in `assertLogs()` in `unittest`.
* Move `TestHandler` from `test.support` to `test_logging`.

https://bugs.python.org/issue40275

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
It depends on #19600 because `asyncio` imports `logging`.

File: Lib/test/test_logging.py
Comment (shihai1991) [inline Lib/test/test_logging.py]: Hm, I am not sure it's only test_logging use this  common class forever.

**Review by vsajip — CHANGES_REQUESTED** [review]
The whole purpose of placing `TestHandler` in `test.support` was that it could be used/useful in other tests. I understand if `test.support` needs to be subdivided further into subpackages, e.g. `test.support.logging` to hold this kind of code, but I don't think it's appropriate to move it into `test_logging` (even if it's the only current user of `TestHandler`) because it then wouldn't be obvious that it was intended for use across any tests that wanted to use logging.

I also see the microbenchmarks on the issue. but what is the practical impact on the time taken for test runs? How much does the total time for a test run reduce as a result of refactoring `test.support` to minimise imports?

**Comment by serhiy-storchaka** [comment]
`TestHandler` was added 9.5 years ago, and it is still only used in a single test in a single file. I am not sure that it will be used in other tests in the next 10 years. But if you prefer, I'll move it into a `test.support` submodule.

The goal of this change is not to save the import time, but to minimize possible side effects of importing the `logging` package. It is relatively complex package which imports 34 modules. Some of them can have side effects which can affect the particular test.

File: Lib/unittest/case.py
Comment (vstinner) [inline Lib/unittest/case.py]: I dislike defining a new _CapturingHandler class each time _AssertLogsContext context manager is used :-( Maybe a lazy import could be used, something similar to PR #19600?

File: Lib/unittest/case.py
Comment (vstinner) [inline Lib/unittest/case.py]: Hum. Would it be possible to move _AssertLogsContext class in a private unittest._log submodule and imports this one on demand?

File: Lib/test/support/logging_helper.py
Comment (vstinner) [inline Lib/test/support/logging_helper.py:3]: It's only used in test_logging. Why not simply moving this class into test_logging?

We don't provide any backward compatibility warranty for test.support. It should not be used outside CPython code base.

**Review by vstinner — COMMENTED** [review]
Maybe moving the logging changes into a separated PR since test_logging/test.support changes and unittest changes are not directly related.

File: Lib/unittest/case.py
Comment (vstinner) [reply Lib/unittest/case.py]: It's kind of unusal to define classes "only once" in a function. A separated submodule imported lazily sounds more common to me.

File: Lib/test/support/logging_helper.py
Comment (serhiy-storchaka) [reply Lib/test/support/logging_helper.py:3]: It is what I did originally. But then move it into a separate submodule on the request of @vsajip (`logging` maintainer, who added this class 10 years ago).

File: Lib/unittest/case.py
Comment (serhiy-storchaka) [reply Lib/unittest/case.py]: Good idea! It looks clearer now.

**Comment by vstinner** [comment]
@vsajip: "The whole purpose of placing TestHandler in test.support was that it could be used/useful in other tests"

The thing is that 10 years later, it's only used by test_logging. If tomorrow, another test wants to use it, support.logging_helper could be added. But I don't think that it's worth it. I would prefer to move TestHandler in test_logging to be able to remove "import logging" from test.support.

@vsajip: "I also see the microbenchmarks on the issue. but what is the practical impact on the time taken for test runs?"

See https://bugs.python.org/issue40275#msg366823

File: Lib/unittest/_log.py
Comment (vstinner) [inline Lib/unittest/_log.py]: Please remove empty lines at the end.

**Review by vstinner — COMMENTED** [review]
Since there is a disagreement on test.support, maybe keep this PR for the unittest module, and write a new one for test.support changes?

**Pull request merged**
